LLVM is statically linked multiple times.

@ptersilie has been experiencing some odd behavior where he makes a change to ykllvm (e.g. adds print statements), recompiles it, then re-runs an interpreter, and he doesn't see his changes take effect.

After showing me this, I had a prod around in the `yk` build system and found some surprising things:

 * We haven't been building shared objects (`.so` files) for LLVM, only static `.a` libraries.
 * no released version of llvm-sys is currently capable of linked llvm dynamically either (the git version is capable).

This means that you can have multiple copies of LLVM inside one yk (why don't symbols clash? no idea), but worse, they may be different versions of LLVM. The latter might explain the problem @ptersilie is experiencing.

We tried to fix it as follows:
 * build shared objects for llvm: add `-DBUILD_SHARED_LIBS=On` when configuring LLVM.
 * use the git head of `llvm-sys` in our `Cargo.toml` files and with the `force-dynamic` feature on.

However, one of our passes then fails to build:

```
/usr/bin/ld: ControlPoint.cpp:(.text._ZN12_GLOBAL__N_114YkControlPoint11runOnModuleERN4llvm6ModuleE+0x1518): undefined reference to `llvm::LoopBase<llvm::BasicBlock, llvm::Loop>::~LoopBase()'
collect2: error: ld returned 1 exit status
```

Probably we need to inspect the cmake dependendencies of our pass.

(we should probably also add `--link-shared` to anywhere in our codebase that uses `llvm-config`)

I'll work on this as soon as I find time. Until then, be aware that you may need to rebuild the whole workspace every time your ykllvm code changes! Not good, I know.

--- a/.buildbot.sh
+++ b/.buildbot.sh
@@ -53,6 +53,7 @@ cmake -DCMAKE_INSTALL_PREFIX=`pwd`/../inst \
     -DLLVM_ENABLE_ASSERTIONS=On \
     -DLLVM_ENABLE_PROJECTS="lld;clang" \
     -DCLANG_DEFAULT_PIE_ON_LINUX=OFF \
+    -DBUILD_SHARED_LIBS=ON \
     -GNinja \
     ../llvm
 cmake --build .

--- a/docs/src/dev/getting_started.md
+++ b/docs/src/dev/getting_started.md
@@ -26,6 +26,7 @@ cmake -DCMAKE_INSTALL_PREFIX=`pwd`/../inst \
     -DCMAKE_BUILD_TYPE=release \
     -DLLVM_ENABLE_ASSERTIONS=On \
     -DLLVM_ENABLE_PROJECTS="lld;clang" \
+    -DBUILD_SHARED_LIBS=ON \
     -DCLANG_DEFAULT_PIE_ON_LINUX=OFF \
     ../llvm
 make -j `nproc` install

--- a/tests/Cargo.toml
+++ b/tests/Cargo.toml
@@ -43,6 +43,9 @@ ykcapi = { path = "../ykcapi", features = ["yk_testing", "yk_jitstate_debug"] }
 ykllvmwrap = { path = "../ykllvmwrap", features = ["yk_testing"] }
 ykrt = { path = "../ykrt", features = ["yk_testing", "yk_jitstate_debug"] }
 
+[build-dependencies]
+ykbuild = { path = "../ykbuild" }
+
 [[bench]]
 name = "bench"
 harness = false

--- a/tests/build.rs
+++ b/tests/build.rs
@@ -1,8 +1,11 @@
 use std::env;
+use ykbuild;
 
 pub fn main() {
     // Expose the cargo profile to run.rs so that it can set the right link flags.
     if let Ok(profile) = env::var("PROFILE") {
         println!("cargo:rustc-cfg=cargo_profile=\"{}\"", profile);
     }
+
+    ykbuild::apply_llvm_ld_library_path();
 }

--- a/ykbuild/src/lib.rs
+++ b/ykbuild/src/lib.rs
@@ -6,6 +6,7 @@ use std::{
     fs::{create_dir_all, File},
     io::{Read, Write},
     path::{Path, PathBuf},
+    process::Command,
 };
 use tempfile::TempDir;
 
@@ -140,4 +141,20 @@ impl CCGenerator {
         write!(outfile, "{}", entries.join(",\n")).unwrap();
         write!(outfile, "\n]\n").unwrap();
     }
+}
+
+pub fn llvm_config() -> Command {
+    let mut c = Command::new("llvm-config");
+    c.arg("--link-shared");
+    c
+}
+
+/// Call from a build script to ensure that the LLVM libraries are in the loader path.
+///
+/// This is preferred to adding an rpath, as we wouldn't want to distribute binaries with
+/// system-local rpaths inside.
+pub fn apply_llvm_ld_library_path() {
+    let lib_dir = llvm_config().arg("--libdir").output().unwrap().stdout;
+    let lib_dir = std::str::from_utf8(&lib_dir).unwrap();
+    println!("cargo:rustc-env=LD_LIBRARY_PATH={}", lib_dir);
 }

--- a/ykcapi/Cargo.toml
+++ b/ykcapi/Cargo.toml
@@ -16,7 +16,17 @@ ykrt = { path = "../ykrt" }
 yktrace = { path = "../yktrace" }
 yksmp = { path = "../yksmp" }
 ykfr = { path = "../ykfr" }
-llvm-sys = "130"
+
+[dependencies.llvm-sys]
+# note: using a git version to get llvm linkage features in llvm-sys (not in a
+# release at the time of writing)
+git = "https://gitlab.com/taricorp/llvm-sys.rs"
+rev = "678b3da2b2239ae12766c964e6e613c0d82b5f37"
+# because yk already links llvm elsewhere.
+features = ["no-llvm-linking"]
+
+[build-dependencies]
+ykbuild = { path = "../ykbuild" }
 
 [features]
 yk_testing = []

--- a/ykcapi/build.rs
+++ b/ykcapi/build.rs
@@ -1,0 +1,5 @@
+use ykbuild;
+
+pub fn main() {
+    ykbuild::apply_llvm_ld_library_path();
+}

--- a/ykcapi/scripts/yk-config
+++ b/ykcapi/scripts/yk-config
@@ -99,7 +99,15 @@ handle_arg() {
             OUTPUT="${OUTPUT} -L${DIR}/../../target/${mode}/deps"
 
             # Encode an rpath so that we don't have to set LD_LIBRARY_PATH.
+            #
+            # FIXME: Adding rpaths should probably be behind a flag. It's kind
+            # of rude to add local rpaths to interpreter binaries that
+            # downstreams may want to distribute.
             OUTPUT="${OUTPUT} -Wl,-rpath=${DIR}/../../target/${mode}/deps"
+            OUTPUT="${OUTPUT} -Wl,-rpath=$(llvm-config --link-shared --libdir)"
+            # Add a proper RPATH, not a RUNPATH:
+            # https://bugs.launchpad.net/ubuntu/+source/glibc/+bug/1737608
+            OUTPUT="${OUTPUT} -Wl,--disable-new-dtags"
 
             # Improve the quality of profiling data.
             OUTPUT="${OUTPUT} -Wl,--no-rosegment"

--- a/ykfr/Cargo.toml
+++ b/ykfr/Cargo.toml
@@ -10,8 +10,15 @@ libc = "0.2.97"
 libffi = "3.0.0"
 ykutil = { path = "../ykutil" }
 yksmp = { path = "../yksmp" }
-llvm-sys = "140"
 memmap2 = "0.5.2"
+
+[dependencies.llvm-sys]
+# note: using a git version to get llvm linkage features in llvm-sys (not in a
+# release at the time of writing)
+git = "https://gitlab.com/taricorp/llvm-sys.rs"
+rev = "678b3da2b2239ae12766c964e6e613c0d82b5f37"
+# because yk already links llvm elsewhere.
+features = ["no-llvm-linking"]
 
 [dependencies.object]
 version = "0.28.3"

--- a/ykllvmwrap/build.rs
+++ b/ykllvmwrap/build.rs
@@ -1,19 +1,15 @@
 #![feature(fn_traits)]
 
 use rerun_except::rerun_except;
-use std::{env, process::Command};
-use ykbuild::{CCGenerator, CCLang};
+use std::env;
+use ykbuild::{apply_llvm_ld_library_path, llvm_config, CCGenerator, CCLang};
 
 fn main() {
     // Ensure changing C++ source files or headers retriggers a build.
     rerun_except(&[]).unwrap();
 
     // Compile our wrappers with the right LLVM C++ flags.
-    let cxxflags_out = Command::new("llvm-config")
-        .arg("--cxxflags")
-        .output()
-        .unwrap()
-        .stdout;
+    let cxxflags_out = llvm_config().arg("--cxxflags").output().unwrap().stdout;
     let cxxflags_str = std::str::from_utf8(&cxxflags_out).unwrap();
     let cxxflags = cxxflags_str.split_whitespace().collect::<Vec<_>>();
 
@@ -58,19 +54,12 @@ fn main() {
     ccg.generate();
 
     // Ensure that downstream crates performing linkage use the right -L and -l flags.
-    let lib_dir = Command::new("llvm-config")
-        .arg("--libdir")
-        .output()
-        .unwrap()
-        .stdout;
+    let lib_dir = llvm_config().arg("--libdir").output().unwrap().stdout;
     let lib_dir = std::str::from_utf8(&lib_dir).unwrap();
     println!("cargo:rustc-link-search={}", lib_dir);
+    apply_llvm_ld_library_path();
 
-    let libs = Command::new("llvm-config")
-        .arg("--libs")
-        .output()
-        .unwrap()
-        .stdout;
+    let libs = llvm_config().arg("--libs").output().unwrap().stdout;
     let libs = std::str::from_utf8(&libs).unwrap();
     for lib in libs.split_whitespace() {
         assert!(lib.starts_with("-l"));

--- a/ykrt/Cargo.toml
+++ b/ykrt/Cargo.toml
@@ -15,6 +15,7 @@ yktrace = { path = "../yktrace" }
 
 [build-dependencies]
 regex = "1.5.4"
+ykbuild = { path = "../ykbuild" }
 
 [features]
 yk_jitstate_debug = []

--- a/ykrt/build.rs
+++ b/ykrt/build.rs
@@ -1,0 +1,5 @@
+use ykbuild;
+
+pub fn main() {
+    ykbuild::apply_llvm_ld_library_path();
+}

